pm.pl

```diff
@@ -1,37 +1,60 @@
 :- use_module(download).
 :- use_module(config).
 :- use_module(find_modules).
-create_dir_if_needed :- \+ exists_directory("deps"),
-     make_directory("deps").
+
+     
+create_dir_if_needed :-   (  exists_directory("deps")
+    -> true
+    ;  make_directory("deps"),
+    true
+).
      
 download_to_disk(URL, Version, PackageName) :- 
+    create_dir_if_needed,
     split_string(URL, "/", "", Pieces),
     last(Pieces, PackageName),
     format(atom(Formatted), 'deps/~w', [PackageName]),
     down(URL, Formatted).   
 
-sync :- create_dir_if_needed,
-        forall(dependency(URL, Version), download_to_disk(URL, Version, _)).
+sync :- 
+    create_dir_if_needed,
+    forall(dependency(URL, Version), download_to_disk(URL, Version, _)).
+
+write_source(Handle, Module, File) :-
+    format(atom(SourceLine), 'source("~w", "~w").', [Module, File]),
+    writeln(Handle, SourceLine).
 
 handle_module_sources(File) :-
-                               format(atom(Formatted), 'deps/~w', [File]),
-                               find_deps(Formatted, [(Module, _)]),
-                               open('deps/sources.pl', append, Handle),
-                               format(atom(SourceLine), 'source("~w", "~w").', [Module, File]),
-                               writeln(Handle, SourceLine),
-                               close(Handle).
+    format(atom(Formatted), 'deps/~w', [File]),
+    (\+ find_deps(Formatted, [(Module, _)])
+    -> print("No modules in file"),
+    delete_file(Formatted),
+    false
+    ; find_deps(Formatted, [(Module, _)]),  
```

Comment:
you have this goal twice (though the first time you're checking whether it _cannot_ be proven). Probably better to reverse the if/then/else structure so that you are trying to prove find_deps to begin with, rather than \+ find_deps

Reply:
I was confused why you're explicitly unifying against a single element list as the 2nd arg here (ie. what happens if we get more than one element?), but I made another comment on the find_deps definition that this should probably never "return" a list as the second arg. At most we should see a single module definition for a single file

Reply:
I like the idea (generally) of handling exceptional cases explicitly so that's what I reached for this flow.

In javascript this would be something like:
```javascript
const moduleDeclarations = getModuleDeclarations();
if (moduleDeclarations.length < 1) return false;
...
```

Maybe that doesn't apply here because we sort of have to have if/else blocks (which the early returning avoids). Any thoughts on alternative ways of writing this?

Reply:
@ZakMiller I still don't understand why you call the predicate twice though. Can't you just call it once as a positive and then, if that fails, do your "error handling" in the else clause of the if/else construct?

Reply:
My issue is that you're basically doing an if-elif where each if has the same condition just inverted